### PR TITLE
array_analysis_SVD.py update.

### PR DESCRIPTION
**Pull request by birdday**
This file was not functioning as is, so some of these commits are not quite atomic - I was just running key lines in atom. That said, I think it's as clear as it can be, but if you have any questions, let me know.

### CODE REVIEW

File: sensor_array/analysis/array_analysis_SVD.py [hunk 1]
```diff
@@ -264,8 +224,7 @@ def calculate_all_arrays_list(mof_list, num_mofs):
         array_size += 1
 
     return mof_array_list
-list_of_arrays = calculate_all_arrays_list(mof_list_filtered, [10,10])
-array = list_of_arrays[0]
+
 
 Sig_list = []
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py]: naming

File: sensor_array/analysis/array_analysis_SVD.py [hunk 2]
```diff
@@ -23,6 +24,76 @@ def yaml_loader(filepath):
...
 
+def load_henrys_data(filepath, gases):
+    dict_of_dataframes = {gas: 0 for gas in gases}
+    for gas in gases:
+        filename = filepath+gas+'.csv'
+        kh_data = pd.read_csv(filename, sep='\t', engine='python')
+        dict_of_dataframes[gas] = kh_data
+
+    return dict_of_dataframes
+
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py]: Couldn't you just do something like this?
```python
def load_henrys_data(path, gases):
    return {gas: pd.read_csv("%s%s.csv" %(path, gas) for gas in gases}
```
Remind me why we need engine='python' here?
Is filepath supposed to include a directory? If so, then this should be `os.path.join(path, "%s.csv" % gas)`

File: sensor_array/analysis/array_analysis_SVD.py [hunk 3]
```diff
@@ -23,6 +24,76 @@ def yaml_loader(filepath):
...
+
+def filter_henrys_data_by_values(dict_of_dataframes, min_comp=0.0):
+    keys = list(dict_of_dataframes.keys())
+    dict_of_dataframes_clean = {key: 0 for key in keys}
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py]:
```python
dict_of_dataframes_clean = {key: 0 for key in dict_of_dataframes.keys()}
```

Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py]: d = {}
for ...
  d[key] = the new thing

Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py]: should be something like this:

df[(df.gas_kh is not NAN) & (df.max_comp >= min_comp]

Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py]: I think this can also be one line!

File: sensor_array/analysis/array_analysis_SVD.py [hunk 4]
```diff
@@ -23,6 +24,76 @@ def yaml_loader(filepath):
...
+    return dict_of_dataframes_clean
+
+
+def find_common_mofs(dict_of_dataframes):
+    keys = list(dict_of_dataframes.keys())
+    all_mofs = []
+    for key in keys:
+        df = dict_of_dataframes[key]
+        mofs = set(df['MOF'])
+        all_mofs.extend([mofs])
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py]: something like this?

```python
all_mofs = set([gas['MOF'].unique() for gas in gases].flatten())

File: sensor_array/analysis/array_analysis_SVD.py [hunk 5]
```diff
@@ -23,6 +24,76 @@ def yaml_loader(filepath):
...
+
+    common_mofs = all_mofs[0]
+    for i in range(1,len(all_mofs)):
+        common_mofs = common_mofs.intersection(all_mofs[i])
+
+    return common_mofs
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py]: or maybe something like this: 
```python
from functools import reduce
reduce(np.intersect1d, ([1, 3, 4, 3], [3, 1, 2, 1], [6, 3, 4, 2]))
array([3])
```

Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py]: or: 

```python
 set.intersection({1, 3, 4, 3}, {3, 1, 2, 1}, {6, 3, 4, 2})
```

File: sensor_array/analysis/array_analysis_SVD.py [hunk 6]
```diff
@@ -23,6 +24,76 @@ def yaml_loader(filepath):
...
+    mass_constants = []
+    for key in keys:
+        df = dict_of_dataframes[key]
+        df_unified[key] = df['gas_kh']
+        mass_constants.extend([list(df['pure_air_mass'])])
+    avg_mass_constants = np.sum(mass_constants, axis=0)/np.shape(mass_constants)[0]
+    df_unified['pure_air_mass'] = avg_mass_constants
+
+    return df_unified
+
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py:29]: similiar treatment to these!

File: sensor_array/analysis/array_analysis_SVD.py [hunk 7]
```diff
@@ -93,13 +93,9 @@ def unify_henrys_data_and_average_constants(dict_of_data_frames):
...
-    
-def calculate_all_arrays_list(mof_list, num_mofs):
-    mof_array_list = []
-    array_size = min(num_mofs)
-    while array_size <= max(num_mofs):
-        mof_array_list.extend(list(combinations(mof_list, array_size)))
-        array_size += 1
+
+def calculate_all_arrays(mof_list, array_size):
+    mof_array_list = list(itertools.combinations(mof_list, array_size))
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py:31]: list()?

File: sensor_array/analysis/array_analysis_SVD.py [hunk 8]
```diff
@@ -100,20 +100,17 @@ def calculate_all_arrays(mof_list, array_size):
...
-def create_henrys_matrix(array, gases, all_kH):
-    H = []
-    for mof in array:
-        temp_row = []
-        for gas in gases:
-            temp_row.extend([all_kH[mof][gas+'_kH']])
-        H.extend([temp_row])
+def create_henrys_matrix(array, gases, kh_dataframe):
+    df = kh_dataframe
+    H = [list(df.loc[df['MOF'].isin(array)][gas]) for gas in gases]
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py]: no caps

Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py]: something like: 
df[df['MOF'] in array]

**Review by paulboone — CHANGES_REQUESTED** [review]
Mostly pandas and simplifying!

File: sensor_array/analysis/array_analysis_SVD.py (hunk 1 above)
Comment (birdday) [reply sensor_array/analysis/array_analysis_SVD.py]: Fixed.

File: sensor_array/analysis/array_analysis_SVD.py (hunk 7 above)
Comment (birdday) [reply sensor_array/analysis/array_analysis_SVD.py:31]: List is necessary to return a second sorted list after evaluation.

File: sensor_array/analysis/array_analysis_SVD.py (hunk 8 above)
Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py]: no caps!
is the .loc necessary?

File: sensor_array/analysis/array_analysis_SVD.py [hunk 9]
```diff
@@ -35,9 +35,9 @@ def calculate_all_arrays(mof_list, array_size):
 
 def create_henrys_matrix(array, gases, kh_dataframe):
     df = kh_dataframe
-    H = [list(df.loc[df['MOF'].isin(array)][gas]) for gas in gases]
+    h_matrix = [list(df.loc[df['MOF'].isin(array)][gas]) for gas in gases]
 
-    return H
+    return h_matrix
```
Comment (paulboone) [inline sensor_array/analysis/array_analysis_SVD.py:40]: ah ok, this should have been squashed. What about that loc though?

**Review by paulboone — CHANGES_REQUESTED** [review]

File: sensor_array/analysis/array_analysis_SVD.py (hunk 9 above)
Comment (birdday) [reply sensor_array/analysis/array_analysis_SVD.py:40]: My bad on forgetting to squash. The ".loc" was unnecessary - deleted in next commit (Can squash if you'd prefer.)

Comment (paulboone) [reply sensor_array/analysis/array_analysis_SVD.py:40]: I'm going to merge this, but can you tell me why you are converting to a list here?

**Comment by paulboone** [comment]
I still want to make some of the pandas code better, but let's keep going!

**Pull request merged**